### PR TITLE
Postgres integration

### PR DESCRIPTION
**Pull request by sabinem**
The PR integrates the repo with the dataset import into a postgres database:

In order to upload a dataset:
- run `statbotData::create_postgres_table(ds)`: it will always overwrite the postgres table. The function also has a dry run option `statbotData::create_postgres_table(ds, dry_run = TRUE)`: with the dry_run option you can check that the datatypes are all correct.
-  Once the dataset is in postgres: set the status in `data/const/statbot_input_data.csv` to 'uploaded'. Then the query testruns will go to the postgres instance. The postgres instance is more strict on the SQL Syntax then the SQLite DB that is used for local test runs
- for the metadata the idea is to collect them all in once with a script and the to also import them into postgres in one go: the script for that is not ready yet, but a draft has been added here: `scripts/collect_metadata.R`

### CODE REVIEW

**Review by cmdoret — COMMENTED** [review]
I would prefer to minimize the required changes on individual datasets.
With the current implementation of `get_data_types_from_metadata`, we would have to include every column in the metadata files.

Since the purpose of this file thus far was to document columns for humans, this is not the case, and they would need extensive changes.

I will try to rewrite `get_data_types_from_metadata` to infer types froim the dataframe

**Comment by sabinem** [comment]
Hello Cyril, you wrote: "With the current implementation of `get_data_types_from_metadata`, we would have to include every column in the metadata files." This is not the case: if you look at the function, it adds the missing fields which are `year` or `jahr` and `spatialunit_uid` by itself. The only needed change is to change the data_types from `numeric` and `categorical` to the actual datatypes used in postgres, maybe that wouldn't be so complicated and also be helpful in the long run.

But of course we can also infer data types from the data frame, so yes if you want to do that then go ahead with this approach. It will be also fine with me.

**Comment by sabinem** [comment]
Hello @cmdoret I went with the proposed approach: only I felt that the numeric fields "NUMERIC" would work better then "DOUBLE PRECISION", see here: https://www.postgresql.org/docs/8.1/datatype.html#DATATYPE-NUMERIC
Therefore I changed all "DOUBLE PRECISION" and replaced them by "NUMERIC", which should even be fine for integers such as `year`.

**Review by cmdoret — APPROVED** [review]

**Comment by sabinem** [comment]
The PR is now finished: 
- I decided against using scripts for the initial set up, but used package functions instead, that are in `R/postgres_db_initial_setup`
- the `app.R` was moved outside of the package and adapted in a way where it shows the remote data as soon as the pipeline has the status `uploaded`. In that case the queries are also run again against the remote instance, as soon as the pipleline gets loaded in the app
- the repo and pipeline workflow are now described in the `README.md`

File: R/testrun_queries.R [hunk 1]
```diff
@@ -130,13 +152,65 @@ log_result <- function(result, output_path) {
...
+#' log_question_and_query(ds$status, output_path)
+#' }
+log_time_and_environment <- function(pipeline_status, output_path) {
+  if (pipeline_status == "uploaded") {
+    env <- "REMOTE"
+  } else {
+    env <- "LOCAL"
+  }
+  sink(file = output_path, append = FALSE)
+  print(paste0("Environment:", env, ", Time:", Sys.time()),
```
Comment (cmdoret) [inline R/testrun_queries.R:171]:
```suggestion
  print(paste0("Environment: ", env, ", Time: ", Sys.time()),
```

File: README.md [hunk 2]
```diff
@@ -1 +1,165 @@
-# statbotData
\ No newline at end of file
+# statbotData
+
+## About
+This repo is part of the [Statbot Swiss Project](https://www.bfs.admin.ch/bfs/de/home/dscc/blog/2023-02-statbot.html)
+
+The project has to parts: 
+1. Generate Training and Test data for the ML Model from Statistical tables belonging to Swiss Fedearal office of Statistics and selected cantonal statistics offices
```
Comment (cmdoret) [inline README.md:7]:
```suggestion
1. Generate Training and Test data for the ML Model from Statistical tables belonging to Swiss Federal office of Statistics and selected cantonal statistics offices
```

File: README.md [hunk 3]
```diff
@@ -1 +1,165 @@
...
+
+This repo relates to the first part 1. above:
+
+## Content of the repo
+
+The repo consists of the following parts:
+- **input data**: in `data/const/statbot_input_data.csv`: imported datasets and spatial units as csv files
+- **package `statbotData`**: the function in the directory `R` make up a package of reusable functions, that are used in the pipeline to import the datasets into postgres 
+- **app**: `app.R` that gives an overview about the repo and the pipelines
+- **pipelines**: in `pipelines`: they contain the dataset processing as it is preapared for the postgres import and the test data for the ML per dataset
```
Comment (cmdoret) [inline README.md:18]:
```suggestion
- **pipelines**: in `pipelines`: they contain the dataset processing as it is preapared for the postgres import and the test data (question / SQL query pairs) for the ML model per dataset
```

File: README.md [hunk 4]
```diff
@@ -1 +1,165 @@
...
+
+## Pipelines
+
+The pipelines are intended to process datasets and the upload them to a remote postgres instance.
```
Comment (cmdoret) [inline README.md:22]:
```suggestion
The pipelines are intended to process datasets and the upload them to a remote postgres instance, for which the credentials can be configured using the `.env` file (`sample.env` is provided as template).
```

File: README.md [hunk 5]
```diff
@@ -1 +1,165 @@
...
+
+Each pipeline in `data/const/statbot_input_data.csv` has the following fields
+- `data_indicator`: id of the pipeline also used as directory name for the pipeline in the repo
+- `status`: `uploaded` means the dataset has been uploaded to postgres, `approved` means the dataset is ready to be uploaded to postgres
+- `lang`: language of the dataset
+
+### Pipeline files
+
+Each pipeline has a directory in `pipelines` by the name of the `data_indicator`. The pipeline script `<data_indicator>.R` downloads the dataset and preprocesses it to produce a tibble: `ds$postgres_export`. This tibble is later uploaded into postgres. 
+The queries in queries.sql` can either run on the this tibble or on the remote postgres instance. Where they run depends on the pipeline status `ds$status`
```
Comment (cmdoret) [inline README.md:34]:
```suggestion
The queries in `queries.sql` can either run on the this tibble or on the remote postgres instance. Where they run depends on the pipeline status `ds$status`
```
md formatting

**Review by cmdoret — CHANGES_REQUESTED** [review]
Awesome work! The code is really clear and this is a great improvement overall. Thanks for the readme as well. I added a few suggestions and small fixes.

**Review by cmdoret — APPROVED** [review]
Looks good to me!

**Pull request merged**